File: groupctl-prototype/pager.py
from enum import IntEnum
from events import (
    Event,
    EventBus,
    EventBusSubscriber,
    LightPadUp,
    PaintTopButton,
    TopButton,
    TopButtonClick,
    PageSelected,
)


class Page(IntEnum):
    groupctl = 0
    project_explorer = 1



class Pager(EventBusSubscriber):
    def __init__(self, bus: EventBus) -> None:
        self.bus = bus
        self.bus.subscribe(self)
        self.current_page: int = 0
        self._repaint()

    def _get_page_count(self):
        # TODO: make calculated
        return 2

    def _prev_page(self):
        if self.current_page > 0:
            self.current_page -= 1
            self._clear()
            self._repaint()
            self.bus.send(PageSelected(self.current_page))
            # FIXME: send PageSelected event
        
    def _next_page(self):
        if self.current_page < self._get_page_count() - 1:
            self.current_page += 1
            self._clear()
            self._repaint()
            self.bus.send(PageSelected(self.current_page))
            # FIXME: send PageSelected event

    def _clear(self):
        for i in range(1, 65):
            self.bus.send(LightPadUp(n=i, color=0))

    def _repaint(self):
        self.bus.send(PaintTopButton(button=TopButton.up, color=0))
        self.bus.send(PaintTopButton(button=TopButton.down, color=0))
        if self.current_page != 0:
            self.bus.send(PaintTopButton(button=TopButton.up, color=69))
        if self.current_page < self._get_page_count() - 1:
            self.bus.send(PaintTopButton(button=TopButton.down, color=69))

    def on(self, event: Event) -> None:
        match event:
            case TopButtonClick(button=TopButton.up):
                self._prev_page()
            case TopButtonClick(button=TopButton.down):
                self._next_page()

File: GroupCtl/src/main/java/dev/tradcode/groupctl/Pager.java
package dev.tradcode.groupctl;

import dev.tradcode.groupctl.events.ClearLaunchpad;
import dev.tradcode.groupctl.events.ClearTwister;
import dev.tradcode.groupctl.events.Event;
import dev.tradcode.groupctl.events.IEventBus;
import dev.tradcode.groupctl.events.IEventBusSubscriber;
import dev.tradcode.groupctl.events.PageSelected;
import dev.tradcode.groupctl.events.PaintTopButton;
import dev.tradcode.groupctl.events.TopButton;
import dev.tradcode.groupctl.events.TopButtonClick;

public class Pager implements IEventBusSubscriber {
    IEventBus bus;
    int currentPage = 0;

    public Pager(IEventBus bus) {
        this.bus = bus;
        this.bus.subscribe(this);
        this.clear();
        this.paint();
    }

    private void prevPage() {
        if (this.currentPage <= 0)
            return;
        this.currentPage--;
        this.clear();
        this.paint();
        this.bus.send(new PageSelected(this.currentPage));
    }

    private void nextPage() {
        if (this.currentPage >= Page.getPageCount() - 1)
            return;
        this.currentPage++;
        this.clear();
        this.paint();
        this.bus.send(new PageSelected(this.currentPage));
    }

    private void clear() {
        this.bus.send(new ClearLaunchpad(), new ClearTwister());
    }

    private void paint() {
        this.bus.send(
            new PaintTopButton(TopButton.UP, 0),
            new PaintTopButton(TopButton.DOWN, 0)
        );
        if (this.currentPage != 0)
            this.bus.send(new PaintTopButton(TopButton.UP, 69));
        if (this.currentPage < Page.getPageCount() - 1)
            this.bus.send(new PaintTopButton(TopButton.DOWN, 69));
    }

    public void on(Event event) {
        switch (event) {
            case TopButtonClick(var btn) when btn == TopButton.UP -> this.prevPage();
            case TopButtonClick(var btn) when btn == TopButton.DOWN -> this.nextPage();
            default -> { }
        }
    }
}
